### PR TITLE
Implement default history management

### PR DESCRIPTION
**Pull request by martinRenou**

### CODE REVIEW

File: include/xeus/xhistory_manager.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#define XEUS_HISTORY_MANAGER_HPP
+
+#include <string>
+#include <vector>
+
+
+namespace xeus
+{
+
+    class xhistory_manager
```
Comment (JohanMabille) [inline include/xeus/xhistory_manager.hpp]:
```cpp
class XEUS_API xhistory_manager
```

Do not forget to include `xeus.hpp`.

File: include/xeus/xin_memory_history_manager.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+****************************************************************************/
+
+#ifndef XEUS_IN_MEMORY_HISTORY_MANAGER_HPP
+#define XEUS_IN_MEMORY_HISTORY_MANAGER_HPP
+
+#include "xhistory_manager.hpp"
+
+namespace xeus
+{
+    class xin_memory_history_manager : public xhistory_manager
```
Comment (JohanMabille) [inline include/xeus/xin_memory_history_manager.hpp]: Same remark here.

**Comment by martinRenou** [comment]
I guess this one is ready :)

File: src/xin_memory_history_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    xjson xin_memory_history_manager::get_tail_impl(int n, bool /*raw*/, bool /*output*/) const
+    {
+        xjson reply;
+        history_type history;
+
+        int i = 0;
+        for (auto it = m_history.cbegin(); it != m_history.cend() && i < n; ++it, ++i)
+        {
+            history.push_back(*it);
+        }
```
Comment (JohanMabille) [inline src/xin_memory_history_manager.cpp]:
```cpp
int count = std::min(n, static_cast<int>(m_history.size()));
std::copy_n(m_history.cbegin(), count, std::back_inserter(history));
```

File: src/xin_memory_history_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    xjson xin_memory_history_manager::get_range_impl(int /*session*/, int start, int stop, bool /*raw*/, bool /*output*/) const
+    {
+        xjson reply;
+        history_type history;
+
+        int i = start;
+        for (auto it = m_history.cbegin(); it != m_history.cend() && i < stop; ++it, ++i)
+        {
+            history.push_back(*it);
+        }
```
Comment (JohanMabille) [inline src/xin_memory_history_manager.cpp]: I think you want to start the copy at `m_history.cbegin() + start`

**Comment by JohanMabille** [comment]
Also can you squash a bit? (maybe not into a single commit, but remove the clean / review commits).

**Comment by martinRenou** [comment]
This needs some tests

File: src/xin_memory_history_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        return reply;
+    }
+
+    xjson xin_memory_history_manager::get_range_impl(int /*session*/, int start, int stop, bool /*raw*/, bool /*output*/) const
+    {
+        xjson reply;
+        history_type history;
+
+        int count = std::min(stop, static_cast<int>(m_history.size())) - start;
```
Comment (JohanMabille) [inline src/xin_memory_history_manager.cpp]: What if `start` is bigger than `stop` and `m_history.size()` ?

Comment (martinRenou) [reply src/xin_memory_history_manager.cpp]: I can raise a `runtime_error` in that case

**Pull request merged**